feat(download): adding a new `Download.cancel` method

starrify opened the pull request:
Download cancellation may be considered useful in these sample scenarios:
- When the user doesn't expect a download to be initiated from the target URL / page at all.
- When the download takes too long.
- When the download has already transmitted / will transmit too much data.

Due to my lack of knowledge with the other browsers, I'm here implementing this feature for only Chromium based on its [`Browser.cancelDownload` CDP command](https://chromedevtools.github.io/devtools-protocol/tot/Browser/#method-cancelDownload).

If that's deemed necessary, I would also be glad to try implementing it for the other browsers in the (near?) future.

A future task may be to further expose download-related information to users, like the expected download size if available (e.g. via [the `totalBytes` parameter of Chromium's `Browser.downloadProgress` event](https://chromedevtools.github.io/devtools-protocol/tot/Browser/#event-downloadProgress)). With such information known, a user may be able to cancel the download as soon as seeing the expected size, other than when the actually received size goes too large.

starrify committed:
feat(download): adding a new `Download.cancel` method

starrify commented:
I've got the subject line edited to add the WIP notice. Please do not merge this PR until a further notice.

From the CI checks I have observed one type of error which I haven't noticed before during local runs ([one example with check run ID 2386035009](https://github.com/microsoft/playwright/pull/6236/checks?check_run_id=2386035009)). I'm able to reproduce the exact issue locally by repeating the tests for many enough times, e.g.:
```
$ time npx folio --config=tests/config/default.config.ts --tag=chromium -g "*cancel*download*" --repeat-each=100
<omitted lines>
  193 passed (2m)
  7 failed
    download.spec.ts:445:3 › [chromium] download event should be able to cancel pending downloads ==
    download.spec.ts:445:3 › [chromium] download event should be able to cancel pending downloads ==
    download.spec.ts:445:3 › [chromium] download event should be able to cancel pending downloads ==
    download.spec.ts:445:3 › [chromium] download event should be able to cancel pending downloads ==
    download.spec.ts:445:3 › [chromium] download event should be able to cancel pending downloads ==
    download.spec.ts:445:3 › [chromium] download event should be able to cancel pending downloads ==
    download.spec.ts:445:3 › [chromium] download event should be able to cancel pending downloads ==
<omitted lines>
```

I'll look into the issue and make appropriate fixes soon.

starrify committed:
test(download): introducing a large payload for the download-delay test

It is observed that Chromium requires a large payload to trigger the
download event early enough, otherwise that might be delayed for several
seconds.
Before the change the "should be able to cancel pending downloads" test
case fails occasionally (e.g. 7 times out of 100 repeats). After the
change a local run of 1000 repeats suceeded with no failure.

starrify commented:
The earlier observed occasional test failure is believed to have been caused Chromium's behavior on downloading: the download event may be delayed for seconds if the initial payload is not large enough.

After having increased the download payload by 4KB (24ca3e4), the test case succeeded in a local run with 1000 repeats:
```
$ npx folio --config=tests/config/default.config.ts --tag=chromium -g "*cancel*download*" --repeat-each=1000

Running 2000 tests using 4 workers
  2000 passed (5m)

  Slow test: download.spec.ts (6m)
```

starrify committed:
test(download): skipping coverage checks on the newly added `download.cancel` API if not Chromium

starrify commented:
Regarding the CI failures (ten in total) on the earlier revision 24ca3e4:

- Four of them ([`ubuntu-18.04 (firefox)`](https://github.com/microsoft/playwright/runs/2386608565), [`ubuntu-20.04 (firefox)`](https://github.com/microsoft/playwright/runs/2386608565), [`ubuntu-18.04 (webkit)`](https://github.com/microsoft/playwright/runs/2386608611), and [`ubuntu-20.04 (webkit)`](https://github.com/microsoft/playwright/runs/2386608642)) failed due to missing coverage of the newly added `download.cancel` API on Firefox / WebKit. I have just added a new commit 0b9105f7 to disable the coverage check for that temporarily, like for the other Chromium-only APIs.
- Another four of them ([`Chrome Stable (Linux)`](https://github.com/microsoft/playwright/runs/2386609726), [`Chrome Stable (Win)`](https://github.com/microsoft/playwright/runs/2386609830), [`Chrome Stable (Mac)`](https://github.com/microsoft/playwright/runs/2386609787), and [`Edge Stable (Win)`](https://github.com/microsoft/playwright/runs/2386609877)) failed because they used a relatively older stable release of Chrome (`89.0.4389.114` observed) while the [`Browser.cancelDownload` CDP method](https://chromedevtools.github.io/devtools-protocol/tot/Browser/#method-cancelDownload) is implemented in [`b449b5c`](https://chromium.googlesource.com/chromium/src/+/b449b5c88408f83d2d66da401f862c144b9c7968), which is first included in revision `91.0.4437.0`. No action is yet taken for these failures.
- The remaining two ([`Headful Linux (chromium)`](https://github.com/microsoft/playwright/runs/2386609356) and [`Android Emulator (shard 2)`](https://github.com/microsoft/playwright/runs/2386609514)) seems rather irrelevant to this PR and are observed in other recent commits as well.

starrify committed:
Merge branch 'master' into add-feature-cancel-download

starrify commented:
Hi @aslushnikov @yury-s! :) Sorry for bothering you. I picked your names by randomly checking a few recently merged PRs and looking for the reviewer's name.

I understand that your team may have been rather busy working on other high priority tasks. However given that I'm rather new to contributing to this project, I'm yet unsure whether there's any other necessary step for me to do before expecting a review.

Could you kindly suggest whether there is something else that I should have done, or shall I just keep waiting patiently? Thanks! :)

yury-s commented on tests/download.spec.ts:
```diff
@@ -31,6 +31,16 @@ it.describe('download event', () => {
+      // Chromium requires a large enough payload to trigger the download event soon enough
+      res.write('a'.repeat(4096));
+      res.write('foo');
+      setTimeout(() => {
```
just flush (see `cork()/uncork()` methods) the response bytes and remove setTimeout altogether so that the transmission never ends (it is cancelled anyway).

yury-s commented on tests/download.spec.ts:
```diff
@@ -433,4 +443,35 @@ it.describe('download event', () => {
   });
+
+  it('should be able to cancel pending downloads', async ({browser, server, browserName}) => {
+    it.skip(browserName !== 'chromium', 'Download cancellation currently implemented for only Chromium');
```
we usually use `it.fixme` for the functionality that is going to be eventually implemented.

yury-s commented on src/server/webkit/wkBrowser.ts:
```diff
@@ -322,4 +322,9 @@ export class WKBrowserContext extends BrowserContext {
+
+  async _doCancelDownload(uuid: string) {
+    // TODO: Have this implemented
+    throw new Error('Download cancellation not yet implemented in Firefox');
```
typo: in WebKit

yury-s commented on src/server/artifact.ts:
```diff
@@ -16,24 +16,27 @@
     super(parent, 'artifact');
     this._localPath = localPath;
     this._unaccessibleErrorMessage = unaccessibleErrorMessage;
+    this._owner = owner;
```
Let's pass `CancellCallback` function instead to limit the information we expose to the Artifact.

yury-s commented on src/server/download.ts:
```diff
@@ -21,14 +21,16 @@ import { Artifact } from './artifact';
-    this.artifact = new Artifact(page, path.join(downloadsPath, uuid), unaccessibleErrorMessage);
+    this.artifact = new Artifact(page, path.join(downloadsPath, uuid), unaccessibleErrorMessage, this);
     this._page = page;
+    this.uuid = uuid;
```
Once you change the artifact to accept `CancelCallback` function this field should not be necessary.

yury-s reviewed (commented):
Looks good in general. Please file a feature request to track all work in this area as we are going to have multiple browser-specific patches. If you feel adventurous you can have a look at implementing this in other browsers too as we'll want to have feature parity across all browsers for this feature.

> * Another four of them ([`Chrome Stable (Linux)`](https://github.com/microsoft/playwright/runs/2386609726), [`Chrome Stable (Win)`](https://github.com/microsoft/playwright/runs/2386609830), [`Chrome Stable (Mac)`](https://github.com/microsoft/playwright/runs/2386609787), and [`Edge Stable (Win)`](https://github.com/microsoft/playwright/runs/2386609877)) failed because they used a relatively older stable release of Chrome

We use `browserVersion` to define expectations in such cases, see [this file](https://github.com/microsoft/playwright/blob/922d9ce1fbb2c0d78f5b214370da4457b7103b94/tests/selectors-misc.spec.ts#L31 
) for example.

starrify committed:
Merge branch 'master' into add-feature-cancel-download

starrify committed:
chore(testing): moving parameter browserVersion from PageEnv to BrowserEnv

starrify committed:
chore(download): multiple minor changes for download.cancel regarding the reviewer feedback:

- Using `response.uncork` in download-related tests to delay the responses.
- Using `it.fixme` than `it.skip` for the download tests that are currently cr-only.
- Fixing typo in `src/server/webkit/wkBrowser.ts`: "in Firefox" -> "in WebKit".
- Redesigning the Download<>Artifact interface for cancelling.
- Allowing the download.cancel test cases only for browser version >=91.

starrify commented:
Hi @yury-s , thanks a lot for your time and the much help opinions!

I have updated this PR to add multiple minor changes:
- Using `response.uncork` in download-related tests to delay the responses.
- Using `it.fixme` than `it.skip` for the download tests that are currently cr-only.
- Fixing typo in `src/server/webkit/wkBrowser.ts`: "in Firefox" -> "in WebKit".
- Redesigning the Download<>Artifact interface for cancelling.
- Allowing the download.cancel test cases only for browser version >=91.

At the latest revision of 56096e8, local tests on my machine failed for only one test case, which is believed to be irrelevant to the new changes:
```
$ time npx folio --config=tests/config/default.config.ts --tag=chromium
<extra lines omitted>
  1780 passed (3m)
  32 skipped
  1 failed
    headful.spec.ts:166:1 › [chromium] focused input should produce the same screenshot ============
<extra lines omitted>
```

Also, repeating the newly test added test cases for 1000 times finished without an error:
```
$ time npx folio --config=tests/config/default.config.ts --tag=chromium -g "*cancel*download*" --repeat-each=1000
<extra lines omitted>
  2000 passed (6m)
<extra lines omitted>
```

Please notice that there's a relatively independent commit e28eb15 which moves the `browserVersion` parameter from `PageEnv` to `BrowserEnv` for testing. Please let me know if you deem this as suitable for a separate branch / PR.

yury-s commented on docs/src/api/class-download.md:
```diff
@@ -54,6 +54,12 @@ downloaded content. If [`option: acceptDownloads`] is not set, download events a
 not performed and user has no access to the downloaded files.
 :::
 
+## async method: Download.cancel
```
Let's call it `Download._cancel` Until it's implemented in all browsers, add a note that this is an experimental API and rename it later to `Download.cancel` when it's supported in all browsers (or just remove it without breaking our API contract if we cannot implement it in other browsers).

yury-s reviewed (commented):
> At the latest revision of [56096e8](https://github.com/microsoft/playwright/commit/56096e87fa7156c6b26b8821b3b2de41e4e56747), local tests on my machine failed for only one test case, which is believed to be irrelevant to the new changes:

Looks like a flake unrelated to your PR.

yury-s commented:
Please file an issue with feature request for Download.cancel and reference it from this PR. This way it'd be easier to keep track of the work related to this functionality.

starrify committed:
Merge branch 'master' into add-feature-cancel-download

starrify committed:
chore(test): add browserVersion to the test arg definition (following e28eb15)

starrify committed:
docs(download): add explicit "experimental" note for Download.cancel

starrify commented:
Hi @yury-s , thanks a lot for the help! :) Please find below updates from my side following your comment:
- I have created feature request #6362 for better tracking the implement state.
- I have updated this PR (currently at aa7752db) to include an explicit note for being experimental for this new method.
- I have **not** got the method renamed (`Download.cancel` -> `Download._cancel`).

I have no objection with your idea of prefixing the method with an underscore for now. I already have the necessary changes made locally.
However right before submitting the changes I noticed that the new naming (`Download._cancel`) would break the documentation task (`npm run doc`), which looks like:

```
$ npm run doc

> playwright-internal@1.11.0-next doc
> node utils/doclint/cli.js

============================
ERROR: missing documentation:
Documented "Download._cancel" not found is sources
============================
```

The reason is believed to be that some existing code would not pick property names that start with an underscore: https://github.com/microsoft/playwright/blob/06a92684f370c4a9df23430dc11fb29c0afaaa5a/utils/doclint/missingDocs.js#L123-L124

I have therefore ignored the renaming task, at least for now. I hereby propose that:
- We leave the current naming of `Download.cancel` as is, if that's not a big problem.
- Or if you still would like it to be `Download._cancel`, I'll go making necessary changes in `utils/doclint/` (possibly other places?) to resolve the issue caused by the leading underscore.

starrify committed:
Merge branch 'master' into add-feature-cancel-download

yury-s commented:
Thanks for filing the issue!

> * Or if you still would like it to be `Download._cancel`, I'll go making necessary changes in `utils/doclint/` (possibly other places as well?) to resolve the issue caused by the leading underscore.

Let's either propagate the docs for the methods starting with underscore and make sure their documentation articulates their experimental status or update linter to not complain about missing docs for experimental methods. We already have a couple of examples of the latter which go under the lint's radar (see e.g. [`_android`](https://github.com/microsoft/playwright/blob/e0aaef5eab1c699b8919c3b025ab1ddcfe87290f/src/client/playwright.ts#L37) property) because they are defined in playwright.ts. Before the implementation is ready for all browsers we cannot make this is a part of the stable public API.

starrify committed:
Merge branch 'master' into add-feature-cancel-download

starrify committed:
chore(download): renaming Download.cancel to Download._cancel

for it being experimental

starrify commented on docs/src/api/class-download.md:
```diff
@@ -54,6 +54,13 @@ downloaded content. If [`option: acceptDownloads`] is not set, download events a
+**Chromium-only** Cancels a download.
+Will not fail if the download is already finished or canceled.
+Upon successful cancellations, `download.failure()` would resolve to `'canceled'`.
+Currently **experimental** and may subject to further changes.
```
Done in bc5b016e :)

yury-s commented on utils/doclint/cli.js:
```diff
@@ -43,6 +43,7 @@ async function run() {
   documentation.filterForLanguage('js');
 
   // Patch README.md
+  /*
```
Revert changes in this file.

starrify replied:
Ah my bad! Sorry that was accidental.. Thanks for observing that! Just pushed a new commit to revert these lines.

yury-s replied:
Looks like this change is still present.

starrify committed:
Reverting the accidentally included changes in utils/doclint/cli.js

starrify committed:
Merge branch 'master' into add-feature-cancel-download

starrify commented:
Just added a new commit 4d4c7a9b for resolving the conflicts.

starrify committed:
Merge branch 'master' into add-feature-cancel-download

starrify commented:
Again merging master into this branch (new commit: ee468a66) to resolve the merging conflicts.

starrify committed:
Fixing the accidentally messed test case "should report downloads with interception"

starrify committed:
Revert "Fixing the accidentally messed test case "should report downloads with interception""

This reverts commit 51582bf875a227df7ede9985e3afe0892658ace6.

starrify committed:
Fixing the accidentally messed test case "should report downloads with interception"

starrify commented:
The last CI run finished with _35 successful and 1 failing checks_. I've got a quick look and [the only failed check](https://github.com/microsoft/playwright/pull/6236/checks?check_run_id=2725571000) seems to be irrelevant to changes in this PR:
- The failed test case seems to be unrelated (which is ["[chromium] download event should close the context without awaiting the failed download"](https://github.com/microsoft/playwright/pull/6236/checks?check_run_id=2725571000#step:8:37))
- The same failure may be observed in other recent CI runs ([one example](https://github.com/microsoft/playwright/runs/2725563292))

starrify committed:
Merge branch 'master' into add-feature-cancel-download

starrify commented:
Again adding a new commit 0138121a to this PR: merging `master` into this branch to resolve the conflicts.

starrify commented:
Glad to know that the change set has been accepted! Thanks a lot Yury and Max for the kind help! :)
Guess this is (finally) the right time for me to start attempting to implement the same feature for Firefox and WebKit..
